### PR TITLE
Allow for custom head method in HTTPEndpoint

### PR DESCRIPTION
**Pull request by theblazehen**
Fix for https://github.com/encode/starlette/issues/1221

### CODE REVIEW

File: starlette/endpoints.py [hunk 1]
```diff
@@ -23,7 +23,15 @@ def __await__(self) -> typing.Generator:
 
     async def dispatch(self) -> None:
         request = Request(self.scope, receive=self.receive)
-        handler_name = "get" if request.method == "HEAD" else request.method.lower()
+
+        if request.method == "HEAD":
+            if hasattr(self, "head"):
+                handler_name = "head"
```
[the remark this answers is not in the document]
Comment (theblazehen) [reply starlette/endpoints.py:29]: No. In which file would the relevant tests be?

[the remark this answers is not in the document]
Comment (aminalaee) [reply starlette/endpoints.py:29]: Not really sure, maybe `tests/test_endpoints.py`. Or any other places using `HTTPEndpoint`

File: starlette/endpoints.py [hunk 2]
```diff
@@ -23,7 +23,15 @@ def __await__(self) -> typing.Generator:
 
     async def dispatch(self) -> None:
         request = Request(self.scope, receive=self.receive)
-        handler_name = "get" if request.method == "HEAD" else request.method.lower()
+
+        if request.method == "HEAD":
```
Comment (aminalaee) [inline starlette/endpoints.py:27]: It's safer to check `request.method.lower() == "head"`

File: starlette/endpoints.py [hunk 3]
```diff
@@ -23,7 +23,15 @@ def __await__(self) -> typing.Generator:
...
+            if hasattr(self, "head"):
+                handler_name = "head"
+            else:
+                handler_name = "get"
+        else:
+            handler_name = request.method.lower()
+
```
Comment (Kludex) [inline starlette/endpoints.py:34]:
```suggestion
        method = request.method.lower()
        handler_name = "get" if method == "head" and not hasattr(self, "head") else method
```

Comment (Kludex) [inline starlette/endpoints.py:34]: We just need to add the `not hasattr(self, "head")` logic. 
```suggestion
        handler_name = (
            "get"
            if request.method == "HEAD" and not hasattr(self, "head")
            else request.method.lower()
        )
```

**Review by Kludex — CHANGES_REQUESTED** [review]
@theblazehen Thanks for the PR! 🎉 

Can you rebase it and check my comment?

**Comment by Kludex** [comment]
@theblazehen Do you wish to resume your work? Otherwise, I'll be closing as stale :kissing:

**Comment by Kludex** [comment]
Close in favor of #1346

**Pull request closed without merging**